x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/aggs/frequentitemsets/ItemSetTraverser.java

```diff
@@ -0,0 +1,131 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.ml.aggs.frequentitemsets;
+
+import org.elasticsearch.core.Releasable;
+import org.elasticsearch.core.Releasables;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Stack;
+
+/**
+ * A traverser that explores the item set tree, so item sets are generated exactly once.
+ *
+ * For example: A tree that builds all combinations would create traverse the set [a, b, c]
+ * several times: a->b->c, a->c->b, b->a->c, b->c->a, c->a->b, c->b->a
+ *
+ * This traverser avoids those duplicates and only traverses [a, b, c] via a->b->c
+ *
+ * With other words: this traverser is only useful if order does not matter ("bag-of-words model").
+ *
+ * Note: In order to avoid churn, the traverser is reusing objects as much as it can,
+ *       see the comments containing the non-optimized code
+ */
+class ItemSetTraverser implements Releasable {
+
+    private final TransactionStore.TopItemIds topItemIds;
+
+    // stack implementation: to avoid object churn this is not implemented as classical stack, but optimized for re-usage
+    // non-optimized: Stack<TransactionStore.TopItemIds.IdIterator> itemIterators = new Stack<>();
+    private List<TransactionStore.TopItemIds.IdIterator> itemIterators = new ArrayList<>();
+    private int stackPosition = 0;
+
+    private Stack<Long> itemIdStack = new Stack<>();
+
+    ItemSetTraverser(TransactionStore.TopItemIds topItemIds) {
+        this.topItemIds = topItemIds;
+
+        // push the first iterator
+        itemIterators.add(topItemIds.iterator());
+    }
+
+    public boolean hasNext() {
+        // check if we are already exhausted
+        // non-optimized: itemIterators.isEmpty()
+        if (stackPosition == -1) {
+            return false;
+        }
+        return itemIterators.get(stackPosition).hasNext();
+    }
+
+    public boolean next() {
+        // check if we are already exhausted
+        // non-optimized: itemIterators.isEmpty()
+        if (stackPosition == -1) {
+            return false;
+        }
+
+        long itemId;
+        for (;;) {
+            if (itemIterators.get(stackPosition).hasNext()) {
+                itemId = itemIterators.get(stackPosition).next();
+
+                // the way the tree is traversed, it should not create dups
+                assert itemIdStack.contains(itemId) == false : "detected duplicate";
```

Comment:
doesn't this iterate the entire stack? Isn't this expensive for an assertion?

Reply:
good question, afaik assertions are disabled in release mode